File: python/ragstack/mcp/backend.py
"""RAGStack HTTP wrapper and result formatting for the MCP server.

Everything here depends only on ``httpx`` — no ``mcp`` SDK — so the tool logic
can be unit-tested with ``httpx.MockTransport`` and no live server. The three
public methods (:meth:`RagStackBackend.search`, :meth:`~.answer`,
:meth:`~.list_collections`) each return a ready-to-display string and never
raise on an API/transport error: they turn failures into a clear message so the
model surfaces something useful instead of a stack trace.
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Any

import httpx

# Truncated snippet length for a retrieved chunk — long enough to judge
# relevance and cite, short enough to keep the tool output scannable.
_SNIPPET_CHARS = 280

# Prefixes the /v1/query endpoint uses for a non-generated (retrieval-only)
# answer — no LLM configured, or generation failed. Kept in sync with
# ``ragstack.api.routers.query._fallback_answer``.
_LLM_NOT_CONFIGURED = "[LLM not configured]"
_GENERATION_FAILED = "[answer generation failed]"


@dataclass(frozen=True)
class RagStackConfig:
    """Connection settings for a RAGStack instance, read from the environment.

    - ``RAGSTACK_BASE_URL`` — base URL of the API (e.g. ``http://localhost:8030``).
    - ``RAGSTACK_API_KEY`` — optional; sent as the ``X-API-Key`` header when set.
    - ``RAGSTACK_COLLECTION`` — optional default collection id for tools that
      accept a ``collection`` argument.
    """

    base_url: str = "http://localhost:8000"
    api_key: str | None = None
    collection: str | None = None

    @classmethod
    def from_env(cls, env: dict[str, str] | None = None) -> RagStackConfig:
        src = env if env is not None else dict(os.environ)
        base = (src.get("RAGSTACK_BASE_URL") or cls.base_url).rstrip("/")
        return cls(
            base_url=base,
            api_key=src.get("RAGSTACK_API_KEY") or None,
            collection=src.get("RAGSTACK_COLLECTION") or None,
        )


def _title_for(meta: dict[str, Any], doc_id: str) -> str:
    """Human label for a source, matching the UI precedence
    (title → filename → source_path → doi → doc_id)."""
    for key in ("title", "filename", "source_path", "doi"):
        val = meta.get(key)
        if val:
            return str(val)
    return doc_id


def _snippet(text: str, limit: int = _SNIPPET_CHARS) -> str:
    flat = " ".join(str(text).split())
    if len(flat) <= limit:
        return flat
    return flat[: limit - 1].rstrip() + "…"


def _format_sources(sources: list[dict[str, Any]]) -> str:
    lines: list[str] = []
    for i, s in enumerate(sources, start=1):
        doc_id = str(s.get("doc_id", "?"))
        chunk_id = str(s.get("chunk_id", "?"))
        meta = s.get("metadata") or {}
        title = _title_for(meta, doc_id)
        score = s.get("score")
        score_str = f"{float(score):.4f}" if isinstance(score, (int, float)) else "n/a"
        lines.append(f"[{i}] {title}  (score {score_str})")
        lines.append(f"    doc_id: {doc_id}  chunk_id: {chunk_id}")
        content = s.get("content")
        if content:
            lines.append(f"    {_snippet(content)}")
    return "\n".join(lines)


class RagStackBackend:
    """Thin async client over the RAGStack HTTP API used by the MCP tools.

    Holds an ``httpx.AsyncClient`` (injectable for tests). Each public method
    returns a display string and handles transport/HTTP errors internally.
    """

    def __init__(self, config: RagStackConfig, http: httpx.AsyncClient) -> None:
        self._config = config
        self._http = http

    @property
    def config(self) -> RagStackConfig:
        return self._config

    def _headers(self) -> dict[str, str]:
        if self._config.api_key:
            return {"X-API-Key": self._config.api_key}
        return {}

    def _resolve_collection(self, collection: str | None) -> str | None:
        return collection or self._config.collection

    async def _request(
        self, method: str, path: str, *, json: dict[str, Any] | None = None
    ) -> tuple[httpx.Response | None, str | None]:
        """Perform a request. Returns ``(response, None)`` on a completed HTTP
        round-trip (any status), or ``(None, message)`` when the server could
        not be reached at all."""
        url = f"{self._config.base_url}{path}"
        try:
            resp = await self._http.request(
                method, url, json=json, headers=self._headers()
            )
            return resp, None
        except httpx.ConnectError:
            return None, (
                f"Cannot reach RAGStack at {self._config.base_url} "
                "(connection refused). Is the server running and is "
                "RAGSTACK_BASE_URL correct?"
            )
        except httpx.HTTPError as exc:
            return None, f"Request to RAGStack at {self._config.base_url} failed: {exc}"

    @staticmethod
    def _http_error_message(resp: httpx.Response) -> str:
        """A concise message for a non-2xx response, pulling the API's
        ``detail`` when present."""
        detail: str | None = None
        try:
            body = resp.json()
            if isinstance(body, dict):
                d = body.get("detail")
                detail = d if isinstance(d, str) else (str(d) if d else None)
        except (ValueError, TypeError):
            detail = None
        if resp.status_code == 401 or resp.status_code == 403:
            return (
                f"RAGStack rejected the request ({resp.status_code}). "
                "Check RAGSTACK_API_KEY."
            )
        if resp.status_code == 503:
            return f"RAGStack is unavailable (503){f': {detail}' if detail else ''}."
        suffix = f": {detail}" if detail else ""
        return f"RAGStack returned HTTP {resp.status_code}{suffix}."

    async def search(
        self, query: str, collection: str | None = None, top_k: int = 5
    ) -> str:
        """Retrieve the top-k chunks for ``query`` via ``POST /v1/retrieve``."""
        if not query.strip():
            return "No query provided. Pass a non-empty 'query'."
        top_k = max(1, min(int(top_k), 50))
        coll = self._resolve_collection(collection)
        payload: dict[str, Any] = {"query": query, "top_k": top_k}
        if coll:
            payload["collection"] = coll

        resp, err = await self._request("POST", "/v1/retrieve", json=payload)
        if err:
            return err
        assert resp is not None
        if resp.status_code >= 400:
            return self._http_error_message(resp)

        try:
            sources = resp.json().get("sources") or []
        except ValueError:
            return "RAGStack returned a malformed response for /v1/retrieve."

        where = f" in collection '{coll}'" if coll else ""
        if not sources:
            return f'No chunks matched "{query}"{where}.'
        header = f'Top {len(sources)} chunks for "{query}"{where}:\n'
        return header + "\n" + _format_sources(sources)

    async def answer(self, query: str, collection: str | None = None) -> str:
        """Full RAG answer for ``query`` via ``POST /v1/query`` (retrieval +
        LLM generation). Degrades to a clear note when no LLM is configured."""
        if not query.strip():
            return "No query provided. Pass a non-empty 'query'."
        coll = self._resolve_collection(collection)
        payload: dict[str, Any] = {"query": query}
        if coll:
            payload["collection"] = coll

        resp, err = await self._request("POST", "/v1/query", json=payload)
        if err:
            return err
        assert resp is not None
        if resp.status_code >= 400:
            return self._http_error_message(resp)

        try:
            body = resp.json()
        except ValueError:
            return "RAGStack returned a malformed response for /v1/query."
        answer_text = str(body.get("answer") or "").strip()
        sources = body.get("sources") or []

        note = ""
        if answer_text.startswith(_LLM_NOT_CONFIGURED):
            note = (
                "Note: no LLM is configured on this RAGStack server, so no "
                "answer was generated. The relevant chunks are listed below — "
                "use them (or the 'search' tool) to answer directly.\n\n"
            )
            answer_text = ""
        elif answer_text.startswith(_GENERATION_FAILED):
            note = (
                "Note: answer generation failed on the RAGStack server; "
                "returning the retrieved chunks only.\n\n"
            )
            answer_text = ""

        parts: list[str] = []
        if note:
            parts.append(note.rstrip())
        if answer_text:
            parts.append(f"Answer:\n{answer_text}")
        if sources:
            parts.append("Sources:\n" + _format_sources(sources))
        elif not answer_text:
            parts.append(f'No relevant chunks were found for "{query}".')
        return "\n\n".join(parts)

    async def list_collections(self) -> str:
        """List queryable collections via ``GET /v1/collections``."""
        resp, err = await self._request("GET", "/v1/collections")
        if err:
            return err
        assert resp is not None
        if resp.status_code >= 400:
            return self._http_error_message(resp)

        try:
            body = resp.json()
        except ValueError:
            return "RAGStack returned a malformed response for /v1/collections."
        collections = body.get("collections") or []
        default = body.get("default")
        if not collections:
            return "This RAGStack instance has no collections registered."

        lines = [f"Available collections (default: {default}):", ""]
        for c in collections:
            cid = c.get("id", "?")
            label = c.get("label") or ""
            model = c.get("model") or "?"
            count = c.get("count")
            text_count = c.get("text_count")
            is_default = " [default]" if c.get("default") else ""
            head = f"- {cid}{is_default}"
            if label:
                head += f'  "{label}"'
            lines.append(head)
            counts = []
            counts.append(
                f"{count} vector chunks" if count is not None else "vector count n/a"
            )
            counts.append(
                f"{text_count} text chunks"
                if text_count is not None
                else "text count n/a"
            )
            lines.append(f"    model: {model};  {', '.join(counts)}")
        return "\n".join(lines)

File: go/internal/mcp/backend.go
// Package mcp implements a standalone Model Context Protocol server that
// exposes a running RAGStack HTTP API to an MCP client (Claude Desktop or
// Claude Code) over stdio.
//
// This file holds the HTTP wrapper and result formatting. It depends only on
// the standard library — no MCP SDK — so the tool logic can be unit-tested
// with net/http/httptest and no live server. The three public methods
// (Backend.Search, Backend.Answer, Backend.ListCollections) each return a
// ready-to-display string and never surface a stack trace: they turn transport
// and API failures into a clear, actionable message. Mirrors the Python server
// in python/ragstack/mcp/backend.py.
package mcp

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"syscall"
	"time"
)

// snippetChars is the truncated snippet length for a retrieved chunk — long
// enough to judge relevance and cite, short enough to keep output scannable.
const snippetChars = 280

// Prefixes the /v1/query endpoint uses for a non-generated (retrieval-only)
// answer — no LLM configured, or generation failed. Kept in sync with
// ragstack.api.routers.query._fallback_answer.
const (
	llmNotConfigured = "[LLM not configured]"
	generationFailed = "[answer generation failed]"
)

// Config holds connection settings for a RAGStack instance, read from the
// environment:
//
//   - RAGSTACK_BASE_URL — base URL of the API (e.g. http://localhost:8030);
//     defaults to http://localhost:8000.
//   - RAGSTACK_API_KEY — optional; sent as the X-API-Key header when set.
//   - RAGSTACK_COLLECTION — optional default collection id for tools that
//     accept a collection argument.
type Config struct {
	BaseURL    string
	APIKey     string
	Collection string
}

// DefaultBaseURL is used when RAGSTACK_BASE_URL is unset or empty.
const DefaultBaseURL = "http://localhost:8000"

// ConfigFromEnv reads the configuration from the given lookup function
// (typically os.Getenv). The base URL is defaulted and stripped of any
// trailing slash so path joining is safe.
func ConfigFromEnv(getenv func(string) string) Config {
	base := getenv("RAGSTACK_BASE_URL")
	if base == "" {
		base = DefaultBaseURL
	}
	base = strings.TrimRight(base, "/")
	return Config{
		BaseURL:    base,
		APIKey:     getenv("RAGSTACK_API_KEY"),
		Collection: getenv("RAGSTACK_COLLECTION"),
	}
}

// Backend is a thin client over the RAGStack HTTP API used by the MCP tools.
// The *http.Client is injectable for tests. Each public method returns a
// display string and handles transport/HTTP errors internally.
type Backend struct {
	cfg    Config
	client *http.Client
}

// NewBackend constructs a Backend. If client is nil a default client with
// sensible timeouts is used.
func NewBackend(cfg Config, client *http.Client) *Backend {
	if client == nil {
		client = &http.Client{Timeout: 60 * time.Second}
	}
	return &Backend{cfg: cfg, client: client}
}

// Config returns the connection settings this backend was built with.
func (b *Backend) Config() Config { return b.cfg }

func (b *Backend) resolveCollection(collection string) string {
	if collection != "" {
		return collection
	}
	return b.cfg.Collection
}

// request performs an HTTP round-trip. It returns the response on any completed
// round-trip (any status code), or a non-empty connErr message when the server
// could not be reached at all. The API key, if any, is never logged.
func (b *Backend) request(
	ctx context.Context, method, path string, body any,
) (resp *http.Response, connErr string) {
	url := b.cfg.BaseURL + path

	var reader io.Reader
	if body != nil {
		buf, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Sprintf("Failed to encode request to RAGStack: %v", err)
		}
		reader = bytes.NewReader(buf)
	}

	req, err := http.NewRequestWithContext(ctx, method, url, reader)
	if err != nil {
		return nil, fmt.Sprintf("Failed to build request to RAGStack at %s: %v", b.cfg.BaseURL, err)
	}
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	if b.cfg.APIKey != "" {
		req.Header.Set("X-API-Key", b.cfg.APIKey)
	}

	resp, err = b.client.Do(req)
	if err != nil {
		if errors.Is(err, syscall.ECONNREFUSED) {
			return nil, fmt.Sprintf(
				"Cannot reach RAGStack at %s (connection refused). "+
					"Is the server running and is RAGSTACK_BASE_URL correct?",
				b.cfg.BaseURL,
			)
		}
		return nil, fmt.Sprintf("Request to RAGStack at %s failed: %v", b.cfg.BaseURL, err)
	}
	return resp, ""
}

// httpErrorMessage builds a concise message for a non-2xx response, pulling the
// API's "detail" field when present.
func httpErrorMessage(status int, raw []byte) string {
	detail := extractDetail(raw)
	switch status {
	case http.StatusUnauthorized, http.StatusForbidden:
		return fmt.Sprintf("RAGStack rejected the request (%d). Check RAGSTACK_API_KEY.", status)
	case http.StatusServiceUnavailable:
		if detail != "" {
			return fmt.Sprintf("RAGStack is unavailable (503): %s.", detail)
		}
		return "RAGStack is unavailable (503)."
	}
	if detail != "" {
		return fmt.Sprintf("RAGStack returned HTTP %d: %s.", status, detail)
	}
	return fmt.Sprintf("RAGStack returned HTTP %d.", status)
}

// extractDetail pulls the "detail" field from a JSON error body, matching the
// Python behaviour: a string is used verbatim, any other truthy value is
// stringified, and an absent/empty value yields "".
func extractDetail(raw []byte) string {
	var body map[string]any
	if err := json.Unmarshal(raw, &body); err != nil {
		return ""
	}
	d, ok := body["detail"]
	if !ok || d == nil {
		return ""
	}
	if s, ok := d.(string); ok {
		return s
	}
	return fmt.Sprint(d)
}

// Search retrieves the top-k chunks for query via POST /v1/retrieve. topK is
// clamped to the range [1, 50] (the caller supplies the default of 5).
func (b *Backend) Search(ctx context.Context, query, collection string, topK int) string {
	if strings.TrimSpace(query) == "" {
		return "No query provided. Pass a non-empty 'query'."
	}
	if topK < 1 {
		topK = 1
	}
	if topK > 50 {
		topK = 50
	}
	coll := b.resolveCollection(collection)

	payload := map[string]any{"query": query, "top_k": topK}
	if coll != "" {
		payload["collection"] = coll
	}

	resp, connErr := b.request(ctx, http.MethodPost, "/v1/retrieve", payload)
	if connErr != "" {
		return connErr
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 400 {
		return httpErrorMessage(resp.StatusCode, raw)
	}

	var body struct {
		Sources []source `json:"sources"`
	}
	if err := json.Unmarshal(raw, &body); err != nil {
		return "RAGStack returned a malformed response for /v1/retrieve."
	}

	where := ""
	if coll != "" {
		where = fmt.Sprintf(" in collection '%s'", coll)
	}
	if len(body.Sources) == 0 {
		return fmt.Sprintf("No chunks matched %q%s.", query, where)
	}
	header := fmt.Sprintf("Top %d chunks for %q%s:\n", len(body.Sources), query, where)
	return header + "\n" + formatSources(body.Sources)
}

// Answer returns a full RAG answer for query via POST /v1/query (retrieval +
// LLM generation). It degrades to a clear note when no LLM is configured or
// generation failed, surfacing the retrieved chunks instead.
func (b *Backend) Answer(ctx context.Context, query, collection string) string {
	if strings.TrimSpace(query) == "" {
		return "No query provided. Pass a non-empty 'query'."
	}
	coll := b.resolveCollection(collection)

	payload := map[string]any{"query": query}
	if coll != "" {
		payload["collection"] = coll
	}

	resp, connErr := b.request(ctx, http.MethodPost, "/v1/query", payload)
	if connErr != "" {
		return connErr
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 400 {
		return httpErrorMessage(resp.StatusCode, raw)
	}

	var body struct {
		Answer  string   `json:"answer"`
		Sources []source `json:"sources"`
	}
	if err := json.Unmarshal(raw, &body); err != nil {
		return "RAGStack returned a malformed response for /v1/query."
	}

	answerText := strings.TrimSpace(body.Answer)
	note := ""
	switch {
	case strings.HasPrefix(answerText, llmNotConfigured):
		note = "Note: no LLM is configured on this RAGStack server, so no " +
			"answer was generated. The relevant chunks are listed below — " +
			"use them (or the 'search' tool) to answer directly."
		answerText = ""
	case strings.HasPrefix(answerText, generationFailed):
		note = "Note: answer generation failed on the RAGStack server; " +
			"returning the retrieved chunks only."
		answerText = ""
	}

	var parts []string
	if note != "" {
		parts = append(parts, note)
	}
	if answerText != "" {
		parts = append(parts, "Answer:\n"+answerText)
	}
	if len(body.Sources) > 0 {
		parts = append(parts, "Sources:\n"+formatSources(body.Sources))
	} else if answerText == "" {
		parts = append(parts, fmt.Sprintf("No relevant chunks were found for %q.", query))
	}
	return strings.Join(parts, "\n\n")
}

// ListCollections lists queryable collections via GET /v1/collections.
func (b *Backend) ListCollections(ctx context.Context) string {
	resp, connErr := b.request(ctx, http.MethodGet, "/v1/collections", nil)
	if connErr != "" {
		return connErr
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 400 {
		return httpErrorMessage(resp.StatusCode, raw)
	}

	var body struct {
		Default     *string      `json:"default"`
		Collections []collection `json:"collections"`
	}
	if err := json.Unmarshal(raw, &body); err != nil {
		return "RAGStack returned a malformed response for /v1/collections."
	}
	if len(body.Collections) == 0 {
		return "This RAGStack instance has no collections registered."
	}

	def := "None"
	if body.Default != nil {
		def = *body.Default
	}
	lines := []string{fmt.Sprintf("Available collections (default: %s):", def), ""}
	for _, c := range body.Collections {
		cid := "?"
		if c.ID != "" {
			cid = c.ID
		}
		head := "- " + cid
		if c.Default {
			head += " [default]"
		}
		if c.Label != "" {
			head += fmt.Sprintf("  %q", c.Label)
		}
		lines = append(lines, head)

		model := "?"
		if c.Model != "" {
			model = c.Model
		}
		var counts string
		if c.Count != nil {
			counts = fmt.Sprintf("%d vector chunks", *c.Count)
		} else {
			counts = "vector count n/a"
		}
		if c.TextCount != nil {
			counts += fmt.Sprintf(", %d text chunks", *c.TextCount)
		} else {
			counts += ", text count n/a"
		}
		lines = append(lines, fmt.Sprintf("    model: %s;  %s", model, counts))
	}
	return strings.Join(lines, "\n")
}

// source is a single retrieved chunk in a /v1/retrieve or /v1/query response.
type source struct {
	DocID    any            `json:"doc_id"`
	ChunkID  any            `json:"chunk_id"`
	Content  string         `json:"content"`
	Score    *float64       `json:"score"`
	Metadata map[string]any `json:"metadata"`
}

// collection is a single entry in a /v1/collections response.
type collection struct {
	ID        string `json:"id"`
	Label     string `json:"label"`
	Model     string `json:"model"`
	Default   bool   `json:"default"`
	Count     *int   `json:"count"`
	TextCount *int   `json:"text_count"`
}

// titleFor returns a human label for a source, matching the UI precedence
// (title → filename → source_path → doi → doc_id).
func titleFor(meta map[string]any, docID string) string {
	for _, key := range []string{"title", "filename", "source_path", "doi"} {
		if v, ok := meta[key]; ok && v != nil {
			if s := fmt.Sprint(v); s != "" {
				return s
			}
		}
	}
	return docID
}

// snippet collapses whitespace in text and truncates to limit runes, appending
// an ellipsis when truncated.
func snippet(text string, limit int) string {
	flat := strings.Join(strings.Fields(text), " ")
	r := []rune(flat)
	if len(r) <= limit {
		return flat
	}
	return strings.TrimRight(string(r[:limit-1]), " ") + "…"
}

func idString(v any) string {
	if v == nil {
		return "?"
	}
	if s, ok := v.(string); ok {
		return s
	}
	// JSON numbers decode to float64; render integers without a trailing ".0".
	if f, ok := v.(float64); ok && f == float64(int64(f)) {
		return fmt.Sprintf("%d", int64(f))
	}
	return fmt.Sprint(v)
}

func formatSources(sources []source) string {
	var lines []string
	for i, s := range sources {
		docID := idString(s.DocID)
		chunkID := idString(s.ChunkID)
		title := titleFor(s.Metadata, docID)
		scoreStr := "n/a"
		if s.Score != nil {
			scoreStr = fmt.Sprintf("%.4f", *s.Score)
		}
		lines = append(lines, fmt.Sprintf("[%d] %s  (score %s)", i+1, title, scoreStr))
		lines = append(lines, fmt.Sprintf("    doc_id: %s  chunk_id: %s", docID, chunkID))
		if s.Content != "" {
			lines = append(lines, "    "+snippet(s.Content, snippetChars))
		}
	}
	return strings.Join(lines, "\n")
}
